### PR TITLE
[Python] improved default argument handling

### PR DESCRIPTION
**Pull request by m7thon**
The current SWIG handling of default arguments for python has several issues:
- Tangled code.
- Incorrect handling of several cases:
  1. Negative octals. Currently not handled correctly when `-py3` given (unusual case, but incorrect).
  2. Arguments of type "octal + something" (e.g. `0640 | 04`). Currently drops everything after the first octal. Nasty!
  3. Bool arguments "0 + something" (e.g. `0 | 1`) are always "False" (unusual case, but incorrect).
- Test cases test only for correct _syntax_, not for correctness of the converted value.
- Different handling of octal literals for Python 2 and Python 3 based on `-py3` switch.

This PR addresses all of the above issues. Here are some comments:
- The relevant conversion code is refactored and should be much more clear now.
- Test cases including some new tests are (slightly) improved (but still only test for corner cases)
- Identical code generated for Python 2 and Python 3. This now writes octals such as `0644` as `int('0644', 8)` (available in Python >= 2.0, propagates to long integer as of 2.3).
- There have always been issues with integer literals exceeding 32 bit on Python <= 2.2. These issues remain.
- Tests now pass with Python 2 and Python 3.

Note that this is a much improved version of the previous PR #699.

### CODE REVIEW

File: Source/Modules/python.cxx [hunk 1]
```diff
@@ -2060,120 +2119,34 @@ class PYTHON:public Language {
...
-    }
-
-    // Check if this is a floating point number (notice that it wasn't
-    // necessarily parsed as a long above, consider e.g. ".123").
-    if (!fail && !result) {
+    result = convertIntegerValue(v, resolved_type);
+    if (!result) {
       result = convertDoubleValue(v);
       if (!result) {
-	if (Strcmp(v, "true") == 0 || Strcmp(v, "TRUE") == 0)
```
Comment (mromberg) [inline Source/Modules/python.cxx]: I realize this commit is recreating the logic (to some extent) of what was already there.  But I wonder about the wisdom of converting "TRUE" and "FALSE" into python's True and False.  

  In C++ "true" and "false" are reserved keywords.  So, doing this for them makes sense.  But "TRUE" and "FALSE" are only possible pre-processor tokens.  And it is possible that they may not be defined in a manner consistent with True and False.

**Comment by mromberg** [comment]
I wonder how well convertIntegerValue() is covered by the existing tests?   It is nice to see more tests with this pull request.  It could be that convertIntegerValue() is used for more than just default arguments.   It would be nice to have tests for all branches inside it.  But I'm not sure how difficult this would be to do.  Just thoughts...

File: Source/Modules/python.cxx (hunk 1 above)
Comment (m7thon) [inline Source/Modules/python.cxx]: Yes, I fully agree. Actually, someone may define FALSE as -1 and use this as a default value for an int, so this could cause a real problem (and it will never do any good).

I'll add a commit to throw those tokens out.

**Comment by m7thon** [comment]
> It could be that convertIntegerValue() is used for more than just default arguments.

No, I checked. That is all it is used for.

> I wonder how well convertIntegerValue() is covered by the existing tests? ... It would be nice to have tests for all branches inside it.

Yes, that would be nice. As I say, the tests are improved, but are far from being meaningful unit tests.

**Comment by mromberg** [comment]
> It could be that convertIntegerValue() is used for more than just default arguments.
> 
> No, I checked. That is all it is used for.

  convertValue() is not declared static.  So, it is (at least in theory) possible to call it from other places other than just python.cxx.  Might be worth sticking a static on there even in your local build tree and make sure no other spots use it.

  If it is only being used for default args then that is nice.  There is at least an indirect way to test it. :)

**Comment by m7thon** [comment]
The `convert...Value()` functions are member functions of `PYTHON` (which inherits from `Language`). Sure they could be made `static` (since they do not depend on the state of the `PYTHON` object), or `private`, or become non-member functions and declared `static` (to be available only in the local compilation unit).

But what's the point? Why should it not be allowed to reuse these functions for some new purpose if someone wants to do so?

**Comment by mromberg** [comment]
> The convert...Value() functions are member functions of PYTHON

  Whoops.  You are right.   I was not used to seeing all those inline functions declared like that in a cxx file.   Disregard my comment.

**Comment by ojwb** [comment]
There are tedious conflicts in `Source/Modules/python.cxx` thanks to a mass whitespace change @wsfulton did since this PR was submitted, so I've squashed it into one commit to make it simpler to reapply.

In pre-C++11 code, technically handling `nullptr` as `NULL` has similar issues to assuming `FALSE` is `0`, but I don't think we have a way to know if we're parsing for C++11 or earlier, and I think the benefits of assuming `nullptr` has its C++11 meaning outweigh the theoretical problems.

Will push once I see the testsuite passing locally.

**Comment by m7thon** [comment]
@ojwb Do you want me to rebase this PR on the current master? I am unsure.

**Comment by ojwb** [comment]
@m7thon Sorry, I meant I'd push once the tests passed, but it seems I got distracted.  Sorry about that.

Have re-rebased and am currently running the tests...

**Comment by ojwb** [comment]
The merged branch failed on a new part of test case `default_args`, because a default parameter such as `unsigned foo = -1U` is converted to constant `-1` in Python which is then rejected because `-1` is out of range for an unsigned.

I've fixed that simply by just handling negative unsigned constants in C/C++ instead:

2425c8d6d896dfb4eacb1b5877a5fa6274fde8d8

More sophisticated would be to convert such constants correctly, but I wanted to get this branch merged - enhancements can be done as a follow-on if someone wants to.
